### ISSUE TITLE
Moodle and Azure

### ISSUE THREAD

**Issue by gitalm** [microsoft/o365-moodle#1897]
Dear Team,

we forgot to update our Client Secret from Microsoft Azure.
Now we are not able to login and change the secret. We fear another lockdown in school and wanted to reactivate Moodle for Teams.

How can i change the secret? 
Thanks for your help

![grafik](https://user-images.githubusercontent.com/7422149/143776285-5f7abe66-e5e0-473f-9590-3a90fcd55891.png)

I found the following recipe from https://support.oneidentity.com/de-de/active-roles/kb/327172/error-aadsts7000222-the-provided-client-secret-keys-are-expired

```
    Log in to the Azure portal at portal.azure.com
    Navigate to Azure Active Directory | App Registrations | Click on your App under Display name (Example: ActiveRoles_AutocreatedAzureBackSyncApp_V2) | Certificates & secrets.
    From here you can see all existing Client Secrets if the above error is seen there will be at least one Secret key showing expired.
    Click on + New client secret, there will be an option to generate a new Client Secret key for this App. Give the Key a description and click Add.
    Once the new Client Secret string is generated, copy this value and save it somewhere.
    Select Azure BackSync connection within the Active Roles Synchronization Service | Connection Settings, and update the Key field with the copied value from the Client Secret string.
    Select Test Connection to verify it passes.
    Select the Save and Continue button to commit the changes.
```

**Comment by weilai-irl** [comment]
Hi @gitalm,

The error itself is self-explanatory - the secret of the Azure app used for Moodle integration has expired. You need to ask your Azure admin to create a new secret, and update the Moodle setting ("Client Secret [auth_oidc | clientsecret]" of auth_oidc) with it.

Regards,
Lai

**Comment by mmulrthelp** [comment]
@gitalm are you after where to change auth_oidc | clientid & auth_oidc | clientsecret within the Moodle database, as you've locked yourself out of the system as the frontend now does not have the correct keys ?
![image](https://user-images.githubusercontent.com/16646247/143847134-c3b2a0ec-a59b-4f34-aa2e-a75dd6710dad.png)

**Comment by gitalm** [comment]
Yes,that's my problem
Can you help?  Thanks in advance

**Comment by mmulrthelp** [comment]
@gitalm I'm not sure where it gets written to on the database and whether its in a readable format. I'll take another look now.

@weilai-irl do you know?

**Comment by mmulrthelp** [comment]
@gitalm I've searched the whole of our 'moodle' schema for the auth_oidc | clientid but it returned nothing.. Have you tried deleting your ids token in moodle.mdl_auth_oidc_token just in case its stored the old one in there? 

Sorry I cannot be of further help. If you do get back in create a local admin account (I did this, just in case)

**Comment by weilai-irl** [comment]
Hi @gitalm,

If you have lost access to the Moodle site as site administrator because of the expired Azure app secret, you can use the following SQL script to reset the secret:
`UPDATE mdl_config_plugins SET value = 'NEW_SECRET' WHERE plugin = 'auth_oidc' AND name = 'clientsecret';`

Regards,
Lai

**Comment by kanekoyuichi** [comment]
Hi @weilai-irl

I have the same issue as @gitalm .
I followed your advice and rewrote the clientsecret, but it still shows the new 'expired'. 
Is there anything else I need to change?

Thanks for your help,
Yuichi

**Comment by weilai-irl** [comment]
Hi @kanekoyuichi,

Providing you generated a new secret and set it correctly using the SQL question (note you need to set the secret value rather than the ID), you should be able to login.

If this doesn't work and your goal is to be able to login to the Moodle site using your site admin account (to troubleshoot), you can temporarily change your site admin account to use "manual" authentication type, and give it a temporary password using SQL. This will allow you login. The SQL queries you will need are:
`
UPDATE mdl_user SET auth = 'manual', password = MD5('TEMPORARY_PASSWORD') WHERE id = 'SITE_ADMIN_USER_ID';
`

Please make sure you change the authentication type back to oidc afterwards. 

In general, it's a good idea to keep a site admin account using manual authentication method.

Regards,
Lai

**Comment by kanekoyuichi** [comment]
Hi @weilai-irl ,

I am grateful for your support.
I followed your advice and changed the authentication type to "manual".
However, when I try to log in 'https://MYDOMAIN.com/login/index.php', I get an error and am redirected to the URL below.
https://MYDOMAIN.com/auth/oidc/

Is there an alternative URL?

Also, 'alternateloginurl' in 'mdl_config' was blank, so I changed it to 'https://MYDOMAIN.com/alternate/login/' using SQL, but when I access it, I get 'Not found'.

Am I doing something wrong?

Best Regards,
Yuichi
